### ISSUE TITLE
eventmachine(1.2.7) installation issue on Intel mac with ruby 3.3.0

### ISSUE THREAD

**Issue by SukeshKonjeti** [eventmachine/eventmachine#990]
**I am trying to upgrade one of my ruby project to 3.3.0 and facing issue like below :**

ERROR:  Error installing eventmachine:
ERROR: Failed to build gem native extension.

    current directory: /Users/sk096170/.rvm/gems/ruby-3.3.0/gems/eventmachine-1.2.7/ext
/Users/sk096170/.rvm/rubies/ruby-3.3.0/bin/ruby extconf.rb --with-ldflags\=-Wl,-undefined,dynamic_lookup
checking for pkg-config for openssl... ["-I/usr/local/Cellar/openssl@1.1/1.1.1w/include ", "-L/usr/local/Cellar/openssl@1.1/1.1.1w/lib", "-lssl -lcrypto"]

Using OpenSSL from pkg-config -I/usr/local/Cellar/openssl@1.1/1.1.1w/include  && -L/usr/local/Cellar/openssl@1.1/1.1.1w/lib && -lssl -lcrypto

checking for -lcrypto... yes
checking for -lssl... yes
checking for openssl/ssl.h... yes
checking for openssl/err.h... yes
checking for rb_trap_immediate in ruby.h,rubysig.h... no
checking for rb_thread_blocking_region()... yes
checking for rb_thread_call_without_gvl() in ruby/thread.h... yes
checking for rb_thread_fd_select()... yes
checking for rb_fdset_t in ruby/intern.h... yes
checking for rb_wait_for_single_fd()... yes
checking for rb_enable_interrupt()... yes
checking for rb_time_new()... yes
checking for inotify_init() in sys/inotify.h... no
checking for __NR_inotify_init in sys/syscall.h... no
checking for writev() in sys/uio.h... yes
checking for pipe2() in unistd.h... yes
checking for accept4() in sys/socket.h... yes
checking for SOCK_CLOEXEC in sys/socket.h... no
checking for sys/event.h... yes
checking for sys/queue.h... yes
checking for clock_gettime()... yes
checking for CLOCK_MONOTONIC_RAW in time.h... yes
checking for CLOCK_MONOTONIC in time.h... yes
CXXFLAGS=-I/usr/local/include -I/usr/local/opt/gcc/include/c++/v1 -fdeclspec
creating Makefile

current directory: /Users/sk096170/.rvm/gems/ruby-3.3.0/gems/eventmachine-1.2.7/ext
make DESTDIR\= sitearchdir\=./.gem.20240829-4094-90sad9 sitelibdir\=./.gem.20240829-4094-90sad9 clean

current directory: /Users/sk096170/.rvm/gems/ruby-3.3.0/gems/eventmachine-1.2.7/ext
make DESTDIR\= sitearchdir\=./.gem.20240829-4094-90sad9 sitelibdir\=./.gem.20240829-4094-90sad9
compiling binder.cpp
make: *** [binder.o] Error 1

make failed, exit code 2

**Tried all below commands but nothing has worked :**
gem install eventmachine -v '1.2.7' -- --with-cppflags=-I/usr/local/include
gem install eventmachine -- --with-openssl-dir=/usr/local/opt/openssl@1.1
gem install eventmachine -v '1.2.7' -- --with-cflags="-Wno-error=implicit-function-declaration"
bundle config set --global build.eventmachine --with-ssl-dir=$(brew --prefix openssl@3)
bundle config build.eventmachine --with-openssl-dir=/opt/homebrew/opt/openssl@1.1
gem install eventmachine -v '1.2.7' -- --with-ldflags="-Wl,-undefined,dynamic_lookup"

Can someone help me to solve this error.

**My Mac details :**
2.6 GHz 6-Core Intel Core i7
Sonoma 14.6.1

**Ruby** 
ruby 3.3.0 (2023-12-25 revision 5124f9ac75) [x86_64-darwin23]

**Comment by jeffbax** [comment]
I'm running into the same with `ruby 3.1.6p260 (2024-05-29 revision a777087be6) [x86_64-darwin24]`

Have tried every combination of those as well, and doesn't seem to be any useful error output to indicate what is really going wrong 💩 

In all the other similar issues with people missing SSL headers for the build it at least said that is what was wrong, whereas now it is just

```
compiling binder.cpp
make: *** [binder.o] Error 1

make failed, exit code 2
```

with... not much to go by

**Comment by sveaspade** [comment]
I'm running in to same with Ruby 3.3.4

```
Fetching gem metadata from http://rubygems.org/.........
Resolving dependencies...
Installing eventmachine 1.2.7 with native extensions
Gem::Ext::BuildError: ERROR: Failed to build gem native extension.
```

```
compiling binder.cpp
make: *** [binder.o] Error 1

make failed, exit code 2
```

**Comment by jeffbax** [comment]
I attempted a full reinstall of Xcode hoping that would spook my machine into compliance but no luck

**Comment by doismellburning** [comment]
I had the same behaviour on my arm64 mac:

```
$ tail -n 5 .rbenv/versions/3.3.5/lib/ruby/gems/3.3.0/extensions/arm64-darwin-23/3.3.0/eventmachine-1.2.7/gem_make.out

make DESTDIR\= sitearchdir\=./.gem.20241009-74780-w0vhdf sitelibdir\=./.gem.20241009-74780-w0vhdf
compiling binder.cpp
make: *** [binder.o] Error 1

make failed, exit code 2
```

Fortunately the Makefile has some convenient options:

```
$ head -n 5 ~/.rbenv/versions/3.3.5/lib/ruby/gems/3.3.0/gems/eventmachine-1.2.7/ext/Makefile

SHELL = /bin/sh

# V=0 quiet, V=1 verbose.  other values don't work.
V = 0
```

And running with `V=1` shows it's running `false`:

```
$ make V=1                                                                                  
false -I. -I$HOME.rbenv/versions/3.3.5/include/ruby-3.3.0/arm64-darwin23 -I$HOME.rbenv/versions/3.3.5/include/ruby-3.3.0/ruby/backward -I$HOME.rbenv/versions/3.3.5/include/ruby-3.3.0 -I. -I/opt/homebrew/Cellar/openssl@3/3.3.2/include -DHAVE_OPENSSL_SSL_H -DHAVE_OPENSSL_ERR_H -DWITH_SSL -DBUILD_FOR_RUBY -DHAVE_RB_THREAD_CALL_WITHOUT_GVL -DHAVE_RB_THREAD_FD_SELECT -DHAVE_TYPE_RB_FDSET_T -DHAVE_RB_WAIT_FOR_SINGLE_FD -DHAVE_RB_TIME_NEW -DHAVE_WRITEV -DOS_UNIX -DHAVE_SYS_EVENT_H -DHAVE_SYS_QUEUE_H -DHAVE_KQUEUE -DOS_DARWIN -DHAVE_CLOCK_GETTIME -DHAVE_CONST_CLOCK_MONOTONIC_RAW -DHAVE_CONST_CLOCK_MONOTONIC -I/opt/homebrew/opt/openssl@3/include -D_XOPEN_SOURCE -D_DARWIN_C_SOURCE -D_DARWIN_UNLIMITED_SELECT -D_REENTRANT   -fno-common -fdeclspec  -arch arm64 -o binder.o -c binder.cpp
make: *** [binder.o] Error 1
```

Inspecting the Makefile more closely:

```
$ grep false Makefile 
CXX = false
```

And that'll do it! (And explains the lack of any helpful message)

I'm not yet sure *why* that's happening and don’t have time to dig further now - I suspect something in my local C++ build environment is not right and autotools thinks I don’t have one - but maybe that helps someone get closer to a solution!

**Comment by liuxr0831** [comment]
I'm having exactly the same issue. Tried different openssl versions. Tried reinstalling command-line developer tools. Nothing worked :(

**Comment by jakehlee** [comment]
> Inspecting the Makefile more closely:
> 
> ```
> $ grep false Makefile 
> CXX = false
> ```
> 
> And that'll do it! (And explains the lack of any helpful message)

Thanks for pointing this out @doismellburning , I also ran into this issue as a novice ruby user (just trying to install Jekyll...) and you pointed me in the right direction. Taking a look at the [extconf.rb](https://github.com/eventmachine/eventmachine/blob/master/ext/extconf.rb), we can see that it relies on `CONFIG["CC"]` and `CONFIG["CXX"]` being set properly. However...

```
% ruby -rrbconfig -e 'puts RbConfig::CONFIG["CC"]'
clang
% ruby -rrbconfig -e 'puts RbConfig::CONFIG["CXX"]'
false
```

This pointed me to a relevant issue thread at https://github.com/rubygems/rubygems/issues/8015

> This is because RVM replaces the `CC` value, if that command is not found at installation time.
> 
> https://github.com/rvm/rvm/blob/master/scripts/functions/build#L68-L87
> 
> While RVM installs `gcc` as a required package and falls back to `"/usr/bin/gcc"`, the binaries are built with `clang`.
> 
> By installing `clang` before `rvm install --binary`, `gem install` worked successfully even without the `CC` environment variable.

So, it's an RVM issue, and it's hopefully easily fixed by

> What worked is fully removing RVM (not just the Rubies) through `rvm implode`, clearing some things it couldn't delete with `sudo rm -rf ~/.rvm`, and reinstalling RVM. I no longer had Clang during all of this, it was able to use the built-in GCC 13.2.

or, on linux,

```
rvm uninstall 3.3.4
rm -rf ~/.rvm/gems/ruby-3.3.4*
rm -rf ~/.rvm/rubies/ruby-3.3.4/

rvm install 3.3.4 --binary
bundle install
```

... however, my issue was on M1 Homebrew with `rbenv`. None of the solutions above worked. All I did was upgrade from Ruby 3.0.0 to 3.3.5. What??

In the end, uninstalling rbenv, homebrew, and xcode and reinstalling everything from scratch worked. 🤦‍♂️ 

https://github.com/brianmario/charlock_holmes/issues/177#issuecomment-2217949483

```
# Uninstall rbenv 
% brew uninstall rbenv ruby-build
% rm -r ~/.rbenv

# Clean out any PATH or source modifications in `.bashrc .bashprofile .zshrc .zprofile`

# Uninstall brew
% /bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/master/uninstall.sh)"
% rm -r /opt/homebrew/

# Uninstall XCode CLT
% sudo rm -rf /Library/Developer/CommandLineTools

# Reinstall XCode CLT
% sudo xcode-select --install
% softwareupdate --all --install --force

# Reinstall brew
% /bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/master/install.sh)"
% brew install openssl@3 readline libyaml gmp autoconf

# Make sure it found CLT and Clang
# Also don't install and link llvm, that also showed up during my research
% brew config
...
Clang: 16.0.0 build 1600
...
CLT: 16.0.0.0.1.1724870825
Xcode: N/A

# Reinstall rbenv
% brew install rbenv ruby-build
% rbenv init

# Install 3.3.5
% rbenv install 3.3.5
ruby-build: using openssl@3 from homebrew
==> Downloading ruby-3.3.5.tar.gz...
-> curl -q -fL -o ruby-3.3.5.tar.gz https://cache.ruby-lang.org/pub/ruby/3.3/ruby-3.3.5.tar.gz
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 21.1M  100 21.1M    0     0  13.3M      0  0:00:01  0:00:01 --:--:-- 13.3M
==> Installing ruby-3.3.5...
ruby-build: using libyaml from homebrew
ruby-build: using gmp from homebrew
-> ./configure "--prefix=$HOME/.rbenv/versions/3.3.5" --with-openssl-dir=/opt/homebrew/opt/openssl@3 --enable-shared --with-libyaml-dir=/opt/homebrew/opt/libyaml --with-gmp-dir=/opt/homebrew/opt/gmp --with-ext=openssl,psych,+
-> make -j 10
-> make install
==> Installed ruby-3.3.5 to /Users/jakelee/.rbenv/versions/3.3.5

NOTE: to activate this Ruby version as the new default, run: rbenv global 3.3.5

% rbenv global 3.3.5
% which ruby
~/.rbenv/shims/ruby

# Confirm it found clang and clang++
% ruby -rrbconfig -e 'puts RbConfig::CONFIG["CC"]'
clang
% ruby -rrbconfig -e 'puts RbConfig::CONFIG["CXX"]'
clang++

# Finally
% gem install eventmachine                         
Fetching eventmachine-1.2.7.gem
Building native extensions. This could take a while...
Successfully installed eventmachine-1.2.7
Parsing documentation for eventmachine-1.2.7
Installing ri documentation for eventmachine-1.2.7
Done installing documentation for eventmachine after 2 seconds
1 gem installed

# Finally finally
% gem install bundler jekyll
...
28 gems installed

# Success!
```

It's worth noting that I knew I had to nuke everything after I saw `CXX` was `false` for `3.1.6`, `3.2.5`, and `3.3.5`.

Good luck folks.

**Comment by jenseickmeyer** [comment]
Thank you for the detailed explanation, @jakehlee. I did re-install Ruby and afterwards I was able to finally install the gem. CLT was already properly installed and configured with Homebrew.

**Comment by jeffbax** [comment]
oh wow, was also using Rbenv and I thought I had a fresh install already -- but I will give it all a shot as well!

the o'l "have you tried turning it off and on again"

*edit*: redoing it all from ground zero worked on my x86, wish I knew how things got into a bad state or why it was only this gem

**Comment by xiaohui-zhangxh** [comment]
@jakehlee  re-install gets eventmachine installed, thanks

**Comment by dblock** [comment]
I ran into this after I had migrated from a non-Mx mac to a new M3 that likely had leftover things with the wrong architecture. Thanks @jakehlee for the deep dive!

tl;dr quick check

```
$ ruby -rrbconfig -e 'puts RbConfig::CONFIG["CXX"]'
false
```

If it says `false`, you have this problem.

I nuked `/Library/Developer/CommandLineTools` and did a `xcode-select --install`, *then* I did a `rvm reinstall 3.3.5 --with-openssl-dir=$(brew --prefix openssl@3)` which picked up the new tools and finally `bundle install` which got me a working eventmachine.

**Comment by amancevice** [comment]
thanks, @dblock this fixed the issue for me, though I use rbenv so my commands were:
```bash
sudo rm -rf /Library/Developer/CommandLineTools
xcode-select --install
rbenv install 3.3.5
ruby -rrbconfig -e 'puts RbConfig::CONFIG["CXX"]' # correctly prints "clang++"
bundle install
```

**Comment by benswift** [comment]
Can confirm this worked for me too (for a ruby installed through `mise`). Previously CXX was being reported as false, and after the reinstall of the xcode tools and uninstall/reinstall of ruby (which mise built from source) the CXX is correctly reported as `clang++`.

Thanks heaps :)

**Comment by teded32** [comment]
Thank you. Worked for me @amancevice.

I am using Rbenv, Ruby 3.0.2, OpenSSL 1.1.1, MacOS Sequoia 15.1 (M1 chip). 

When re-installing Ruby 3.0.2, I had to use a patch explained in this [thread](https://github.com/rbenv/ruby-build/discussions/2185?sort=new#discussioncomment-10756658)

**Comment by jettjaniak** [comment]
Thanks @jakehlee
I was using chruby according to [Jekyll's guide](https://jekyllrb.com/docs/installation/macos/) on my Mac M1.
Running `ruby -rrbconfig -e 'puts RbConfig::CONFIG["CXX"]'` initially showed `false`, but after reinstalling XCode CLT and ruby it shows `clang++` now and Jekyll and eventmachine installed successfully. I didn't have to reinstall homebrew.

**Comment by ymnandan** [comment]
https://mac.install.guide/commandlinetools/6 - to uninstall
and
https://mac.install.guide/commandlinetools/7 - to reinstall

ruby -rrbconfig -e 'puts RbConfig::CONFIG["CXX"]' showed clang++

gem install eventmachine    -- worked

% gem install eventmachine                        
Building native extensions. This could take a while...
Successfully installed eventmachine-1.2.7
Parsing documentation for eventmachine-1.2.7
Installing ri documentation for eventmachine-1.2.7
Done installing documentation for eventmachine after 2 seconds
1 gem installed

**Comment by heathwhughes** [comment]
@amancevice @jakehlee Thank you guys so much! I'm using asdf for ruby installation, but this worked perfectly for me as well!

**Comment by cuteshaun** [comment]
Thanks @jakehlee and @amancevice I was stuck with eventmachine failing to build (`make: *** [binder.o] Error 1`) and found that the cause was:
`ruby -rrbconfig -e 'puts RbConfig::CONFIG["CXX"]'  => false` means we can't work with eventmachine

_Let me summarize a solution:

## How to install Event Machine.

### System info:
- MacBook Pro M4 Pro, macOS Sequoia 15.15
- Ruby 3.1.4 via rbenv

### To fix it, I did:
1. sudo rm -rf /Library/Developer/CommandLineTools
2. xcode-select --install
3. softwareupdate --list
4. sudo softwareupdate --install "Command Line Tools for Xcode-16.4"
5. gem install eventmachine 

### Verified it was fixed:
`ruby -rrbconfig -e 'puts RbConfig::CONFIG["CXX"]'  => clang++` means all good and native extensions can build again

**Comment by facundopadilla** [comment]
the solutions from not working for me in mac m2 with sequoia 15.5, any else with the same problem? :(

**Comment by kaskavalci** [comment]
@facundopadilla I had the same issue and this successfully solved my problem https://github.com/eventmachine/eventmachine/issues/1010#issuecomment-2535473656

tldr
- reinstall xcode-select 
- i've installed xcode and accepted the license (not sure if that's needed)
- reinstall ruby
- ensure that `ruby -rmkmf -e 'puts RbConfig::CONFIG["CXX"]'` returns `clang++`
- install `eventmachine`
